backend/capellacollab/cli/ws.py

```diff
@@ -0,0 +1,355 @@
+    v1 = client.CoreV1Api()
+    targz = out / f"{volume_name}.tar.gz"
+
+    with pod_for_volume(volume_name, namespace, MOUNT_PATH, v1) as pod_name:
+        print(f"Downloading workspace volume to '{targz}'")
+
+        with targz.open("wb") as outfile:
+            for data in stream_tar_from_pod(
+                pod_name, namespace, ["tar", "zcf", "-", MOUNT_PATH], v1
+            ):
+                outfile.write(data)
+
+
+@app.command()
+def restore(
+    volume_name: str,
+    tarfile: Annotated[Path, typer.Argument(exists=True)],
+    namespace: Annotated[str, NamespaceOption],
+    access_mode: str = "ReadWriteMany",
+    storage_class_name: str = "persistent-sessions-csi",
+    user_id: str | None = None,
+):
+    """Restore a backup to a Kubernetes Persistent Volume.
+
+    If the volume does not exist, it is created. Note that
+    `access-mode` and `storage-class-name` should match settings
+    for your cluster.
+
+    Optionally a `user-id` can be provided. All files then be owned
+    by this user id.
+    """
+    v1 = client.CoreV1Api()
+
+    create_persistent_volume(
+        volume_name, namespace, access_mode, storage_class_name, v1
+    )
+
+    with pod_for_volume(
+        volume_name, namespace, MOUNT_PATH, v1, read_only=False
+    ) as pod_name:
+        print(f"Restoring workspace volume to '{volume_name}'")
+
+        with tarfile.open("rb") as infile:
+            stream_tar_to_pod(pod_name, namespace, infile, v1)
+
+        adjust_directory_permissions(
+            pod_name,
+            namespace,
+            v1,
+            user_id,
+        )
+
+
+@contextlib.contextmanager
+def pod_for_volume(
+    volume_name: str,
+    namespace: str,
+    mount_path: str,
+    v1: client.CoreV1Api,
+    read_only=True,
+):
+    name = f"ws-download-{volume_name}-{uuid.uuid1()}"[:63]
+
+    containers = [
+        client.V1Container(
+            name=name,
+            image="alpine:latest",
+            command=["sleep", "infinity"],
+            volume_mounts=[
+                client.V1VolumeMount(
+                    name="vol",
+                    mount_path=mount_path,
+                    read_only=read_only,
+                )
+            ],
+            image_pull_policy="Always",
+        )
+    ]
+
+    volumes_ = [
+        client.V1Volume(
+            name="vol",
+            persistent_volume_claim=client.V1PersistentVolumeClaimVolumeSource(
+                claim_name=volume_name
+            ),
+        )
+    ]
+
+    pod = client.V1Pod(
+        kind="Pod",
+        api_version="v1",
+        metadata=client.V1ObjectMeta(name=name),
+        spec=client.V1PodSpec(
+            containers=containers,
+            volumes=volumes_,
+            restart_policy="Never",
+        ),
+    )
+
+    v1.create_namespaced_pod(namespace, pod)
+
+    while not is_pod_ready(name, namespace, v1):
+        print("Waiting for pod to come online...")
+        time.sleep(2)
```

Comment:
I am not sure whether we should include a timeout here. Obviously, the user can just cancel the command but there are certainly cases where either the users forgets about the command (when there is a long runtime) or is not sure whether he can actually cancel it due to a problem. I would probably suggest a timeout of 3-5 minutes.
What do you think about this?

Reply:
Yup, a little timeout would be nice.